### ISSUE TITLE
unexpected lint warning when providing mutiple files

### ISSUE THREAD

**Issue by xinnjie** [googleapis/api-linter#1465]
#### Environment details

```
$ api-linter --version                                                                                                                                                                                  [22:41:48]
api-linter 1.68.0

$ tree proto                                                                                                                                                                                            [22:44:30]
proto
├── google
│   └── api
│       ├── annotations.proto
│       └── http.proto
└── ....
```

#### Steps to reproduce

1. `api-linter` works great when only provided with one file
```
$ api-linter -I proto proto/google/api/http.proto                                                                                                                                                       [22:45:42]
- file_path: proto/google/api/http.proto
  problems:
    - message: 'Proto package and directory structure mismatch: The proto package must match the proto directory structure.'
      location:
        start_position:
            line_number: 17
            column_number: 1
        end_position:
            line_number: 17
            column_number: 19
        path: proto/google/api/http.proto
      rule_id: core::0191::proto-package
      rule_doc_uri: https://linter.aip.dev/191/proto-package
    ....
```

```
$ api-linter -I proto proto/google/api/annotations.proto                                                                                                                                                [22:47:37]
- file_path: proto/google/api/annotations.proto
  problems:
    - message: 'Proto package and directory structure mismatch: The proto package must match the proto directory structure.'
      location:
        start_position:
            line_number: 17
            column_number: 1
        end_position:
            line_number: 17
            column_number: 19
        path: proto/google/api/annotations.proto
      rule_id: core::0191::proto-package
      rule_doc_uri: https://linter.aip.dev/191/proto-package
```
3. Having unexpected lint warning when providing mutiple files
```
$ api-linter -I proto proto/google/api/http.proto proto/google/api/annotations.proto                                                                                                                    [22:44:53]
2025/01/31 22:45:42 google/api/http.proto:29:9: symbol "google.api.Http" already defined at proto/google/api/http.proto:29:9
google/api/http.proto:33:21: symbol "google.api.Http.rules" already defined at proto/google/api/http.proto:33:21
google/api/http.proto:41:8: symbol "google.api.Http.fully_decode_reserved_expansion" already defined at proto/google/api/http.proto:41:8
google/api/http.proto:308:9: symbol "google.api.HttpRule" already defined at proto/google/api/http.proto:308:9
google/api/http.proto:313:10: symbol "google.api.HttpRule.selector" already defined at proto/google/api/http.proto:313:10
google/api/http.proto:321:12: symbol "google.api.HttpRule.get" already defined at proto/google/api/http.proto:321:12
google/api/http.proto:324:12: symbol "google.api.HttpRule.put" already defined at proto/google/api/http.proto:324:12
google/api/http.proto:327:12: symbol "google.api.HttpRule.post" already defined at proto/google/api/http.proto:327:12
google/api/http.proto:330:12: symbol "google.api.HttpRule.delete" already defined at proto/google/api/http.proto:330:12
google/api/http.proto:333:12: symbol "google.api.HttpRule.patch" already defined at proto/google/api/http.proto:333:12
google/api/http.proto:339:23: symbol "google.api.HttpRule.custom" already defined at proto/google/api/http.proto:339:23
google/api/http.proto:348:10: symbol "google.api.HttpRule.body" already defined at proto/google/api/http.proto:348:10
google/api/http.proto:356:10: symbol "google.api.HttpRule.response_body" already defined at proto/google/api/http.proto:356:10
google/api/http.proto:361:21: symbol "google.api.HttpRule.additional_bindings" already defined at proto/google/api/http.proto:361:21
google/api/http.proto:318:9: symbol "google.api.HttpRule.pattern" already defined at proto/google/api/http.proto:318:9
google/api/http.proto:365:9: symbol "google.api.CustomHttpPattern" already defined at proto/google/api/http.proto:365:9
google/api/http.proto:367:10: symbol "google.api.CustomHttpPattern.kind" already defined at proto/google/api/http.proto:367:10
google/api/http.proto:370:10: symbol "google.api.CustomHttpPattern.path" already defined at proto/google/api/http.proto:370:10
```

This behavior is weird and looks like a bug.

**Comment by noahdietz** [comment]
Hey @xinnjie thanks for the report - I can reproduce this. I think it's related to something in the relationship between `http.proto` and `annotations.proto` - the former is imported by the latter. I tested this theory with another package, where one file imported the other, and the same issue arose. All of the files appear in the `-I` import paths, but A is imported by B AND being "parsed" as input - I think this results in duplicate parsing/registration in proto parse.

**Comment by noahdietz** [comment]
Interesting update - I supplied absolute paths for all inputs (`-I` value and the protos-to-lint) and it worked without error.

Similarly, I supplied just the `-I` imports value as an absolute path, but the protos-to-linter as `-I`-relative values, and it worked without error.

Where we pumped into the issue is when we supplied a relative path for the `-I` imports value and relative protos-to-lint.

The area of interest in `api-linter` is here: https://github.com/googleapis/api-linter/blob/ecc28f0fc9e3a7ed4dbd76d38540224d8e8c8427/cmd/api-linter/cli.go#L153-L172

I think [`protoparse.ResolveFilenames`](https://pkg.go.dev/github.com/jhump/protoreflect@v1.16.0/desc/protoparse#Parser.ParseFiles) may have a part to play here as it has specific logic for handling absolute vs. relative file paths.

Continuing the search

**Comment by noahdietz** [comment]
I updated too quickly - one more test produced a further narrowing result: Running from the shared "root" directory of the protos worked without error:
`api-linter -I .  google/api/annotations.proto google/api/http.proto`

So to narrow it down: The issue appears when `api-linter` is run from above the protos, and the "shared root" in `-I` and the protos-to-lint are both supplied as project-relative paths.

Now - if you supply the `-I` as project-relative path, but the protos-to-lint as **"shared root" relative** it works without error:
`api-linter -I protos  google/api/annotations.proto google/api/http.proto`

This definitely makes me think `protoparse.ResolveFilenames` is in play here.

**Comment by noahdietz** [comment]
I think this might actually be an issue with `api-linter` always appending `.` to the list of imports! https://github.com/googleapis/api-linter/blob/ecc28f0fc9e3a7ed4dbd76d38540224d8e8c8427/cmd/api-linter/cli.go#L97

It means that the protos parsed from `-I proto` and `-I .` are duplicated when the project-relative `proto/....proto` files are parsed as input proto files but their `import` statements use the "shared root"-relative path in the proto.

Basically, when a `-I` values that is relative to `.` is provided, we need to dedupe that/not provide `.` somehow.

**Comment by noahdietz** [comment]
One idea is to move the appending of `"."` to the `protoparse.Parser` init so that it doesn't interfere with the `protoparse.ResolveFilenames`. Testing this out, it seems to work well, because the protos-to-lint are appropriately resolved relative to the shared prefix provided by `-I` that `.` was interfering with, and file names do not need to be resolved against `"."` because there is nothing to resolve them against.

**Comment by xinnjie** [comment]
@noahdietz Thanks for your detailed explanation.

**Comment by xinnjie** [comment]
Though we should also consider if this will break other use cases.

**Comment by noahdietz** [comment]
> Though we should also consider if this will break other use cases.

I've been thinking about that, and spoke to it a bit in the #1466 description. For `ResolveFilenames`, if any of the paths-to-resolve are deemed relative to at _least one_ of the provided paths-to-resolve-**against**, the path-to-resolve is unchanged. This was the problem with including `"."` in the paths-to-resolve-against, it made any relative path-to-resolve automatically unresolved against the other paths-resolve-against. And when you think about it, a relative path provided to a CLI is typically relative to `"."` already.

I think we should be OK.

**Comment by fredrikaverpil** [comment]
This was a breaking change. I'm using the `--descriptor-set-in` argument and now have to set the `--proto-path=.` or something else to keep making api-linter work.

**Comment by noahdietz** [comment]
@fredrikaverpil what version of ApiLinter were you using? Because I shipped a separate fix in v1.69.1 for https://github.com/googleapis/api-linter/issues/1471. If you could provide a rough description of the proto file layout as it appears in the descriptor set and the `api-linter` incantation you used so that I may repro, it would be really useful. Thanks!

**Comment by fredrikaverpil** [comment]
- We bumped from v1.67.2 to v1.69.1 and noticed this change.
- Have a look at [this proto](https://github.com/fredrikaverpil/go-microservice/blob/main/proto/gomicroservice/v1/user_service.proto#L90)
- Make a change, like changing `create_time` to IMMUTABLE or something which shouldn't be valid.
- Run: `cd proto && buf build -o descriptor`
- Run: `cd proto && api-linter --descriptor-set-in=descriptor`

**Comment by noahdietz** [comment]
Ok great, that's useful thank you. And the error message you got?

**Comment by noahdietz** [comment]
For posterity, my initial test is as follows:
```bash
cd $GOOGLEAPIS
protoc -o ~/protos.desc --include_imports -I. google/api/http.proto google/api/annotations.proto
cd ~/

# with leading `./`
api-linter --descriptor-set-in=protos.desc ./google/api/annotations.proto

# without leading `./`
api-linter --descriptor-set-in=protos.desc google/api/annotations.proto

# with imported proto as input (original source of this issue)
api-linter --descriptor-set-in=protos.desc google/api/annotations.proto google/api/http.proto
```

All seem to run OK.

Not saying there isn't an issue, just sharing what I've tried.

**Comment by fredrikaverpil** [comment]
Yeah I realized I actually can't repro this on the commandline. But when I run the linter inside Neovim it doesn't work at all with v1.69.1 (no output received at all from api-linter). It's getting late here but I'll see if it has anything to do with how the plugin receives the results from the execution. [Here's the source](https://github.com/fredrikaverpil/dotfiles/blob/92d9a1190b0cf2a7041b2724062e818da3400acb/nvim-fredrik/lua/fredrik/plugins/lang/protobuf.lua#L139) in case you are interested in digging in. I can switch back and forth between api-linter versions and reliably reproduce this.


Originally, a co-worker realized the problem on their end when using a library which wraps api-linter. Here's their PR to fix it: https://github.com/einride/sage/pull/642
I bumped api-linter just one commit earlier but hadn't noticed any problems at that time: https://github.com/einride/sage/pull/640

**Comment by noahdietz** [comment]
Thank you @fredrikaverpil for all of the testing! LMK what the error message is when you find it. I will try to take a look at the links you provided.

**Comment by fredrikaverpil** [comment]
@noahdietz so in Neovim, I managed to extract this from stderr when on v1.69.1:

```
2025/02/19 19:49:25 must specify at least one import path if any absolute file paths are given
Linter command `api-linter` exited with code: 1
```

If I add `--proto-path=.`, there's no error.

**Comment by noahdietz** [comment]
This is *very* helpful thank you @fredrikaverpil . That points to the exact behavior of `protoparse.ResolveFilenames` that needs to be handled.

**Comment by fredrikaverpil** [comment]
Just for completeness, this is the exact arguments passed in to `uv.spawn` (which executes api-linter):


```lua
{
  args = { "--output-format=json", "--disable-rule=core::0191::java-multiple-files", "--disable-rule=core::0191::java-package", "--disable-rule=core::0191::java-outer-classname", "--descriptor-set-in=/tmp/lua_YltFxU", "/Users/fredrik/code/public/go-microservice/proto/gomicroservice/v1/user_service.proto" },
  cwd = "/Users/fredrik/code/public/go-microservice/proto",
  detached = true,
  stdio = { <userdata 1>, <userdata 2>, <userdata 3> }
}
```

**Comment by noahdietz** [comment]
Ok so I see what I did incorrectly: If no `--proto-path` are specified, we want only want to attempt resolving against `.`.

We still don't want to supply both `--proto-path` inputs and `.` together to `protoparse.ResolveFilenames` because it may result in an improperly resolved file name and then duplicate proto entry as originally mentioned in this issue.

**Comment by fredrikaverpil** [comment]
@noahdietz additional context: https://github.com/einride/sage/pull/642#issuecomment-2669605889

**Comment by noahdietz** [comment]
Also filed https://github.com/googleapis/api-linter/issues/1479 to come up with a test harness for the CLI. Local testing was not good enough here and we can do better.

**Comment by noahdietz** [comment]
@fredrikaverpil let's try this again, shall we? v1.69.2 contains the fix.

When you get a chance, please verify that it works for you. I've tested various absolute path inputs and I believe it will do the job, even with your `--proto-path=.` patch still in place

**Comment by fredrikaverpil** [comment]
@noahdietz nice! It appears as if this is fine now with v1.69.2 ❤

**Comment by noahdietz** [comment]
> [@noahdietz[<img alt="" width="15" height="15" src="chrome-extension://dlebflppeeemcdpidccbiblndppbmjmh/ospo-chrome-ext-logo.png">](https://teams.googleplex.com/ndietz@google.com)](https://github.com/noahdietz) nice! It appears as if this is fine now with v1.69.2 ❤ I can remove `--proto-path=.` and not get any errors/issues.

woohoo!! Thanks so much for the responsiveness after hours on your side, and the patience while I got the fix going.

**Comment by noahdietz** [comment]
@fredrikaverpil our previous fix was not as robust as previously believed so @quirogas worked on a more robust fix intended to address the duplication created by the original issue of supplying a subdirectory of the CWD as an `-I` explicit import. We've been testing it in a variety of cases and with `descriptor-set-in` sans any `-I` and we think it is working. If you are interested in trying it out in your neovim set up to be extra sure, please do so when it is merged if you are able to install a commit rather than a release. We aren't in a rush to release as it is but will likely do so next week.

**Comment by fredrikaverpil** [comment]
@noahdietz @quirogas I can confirm that everything works just fine on my end, in Neovim, when using a build from the main branch, commit [7ecaa42](https://github.com/googleapis/api-linter/commit/7ecaa4200da7b5cbcbf1c273fc77d524f346ae1c) 😄👍

**Comment by noahdietz** [comment]
Awesome thank you so much @fredrikaverpil for checking so quickly :)
